test/e2e/upgrades/persistent_volumes.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+// PersistentVolumeUpgradeTest test that a pv is available before and after a cluster upgrade.
+type PersistentVolumeUpgradeTest struct {
+	pvSource *v1.PersistentVolumeSource
+	pv       *v1.PersistentVolume
+	pvc      *v1.PersistentVolumeClaim
+}
+
+const (
+	PVTestFile string = "/mnt/pv_upgrade_test"
+	PVTestData string = "keep it pv"
+	PVWriteCmd string = "echo \"" + PVTestData + "\" > " + PVTestFile
+	PVReadCmd  string = "cat " + PVTestFile
+)
+
+func (t *PersistentVolumeUpgradeTest) createGCEVolume() *v1.PersistentVolumeSource {
+	diskName, err := framework.CreatePDWithRetry()
+	framework.ExpectNoError(err)
+	return &v1.PersistentVolumeSource{
+		GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+			PDName:   diskName,
+			FSType:   "ext3",
+			ReadOnly: false,
+		},
+	}
+}
+func (t *PersistentVolumeUpgradeTest) deleteGCEVolume(pvSource *v1.PersistentVolumeSource) {
+	framework.DeletePDWithRetry(pvSource.GCEPersistentDisk.PDName)
+}
+
+// Setup creates a pv and then verifies that a pod can consume it.  The pod writes data to the volume.
+func (t *PersistentVolumeUpgradeTest) Setup(f *framework.Framework) {
+
+	// TODO: generalize this to other providers
+	framework.SkipUnlessProviderIs("gce")
```

Comment:
Could this also work with GKE? If not, we have to still run this test manually for GKE.